subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClassLoaderUtils.java

```diff
@@ -89,6 +106,51 @@ static Package getPackage(ClassLoader classLoader, String name) {
         Package getPackage(ClassLoader classLoader, String name);
     }
 
+    /**
+     * This class makes it a bit easier to use {@link MethodHandles.Lookup}.
+     * In order to access a method, a lookup object which is accessible to this method must be provided.
+     * Usually, this class and the target Gradle-managed class loader exist in the same module, so everything works.
+     * Otherwise, an {@link IllegalAccessException} will be thrown, and {@link ClassLoader} class will be used as the lookup object.
+     */
+    private static class AbstractClassLoaderLookuper {
+        protected MethodHandles.Lookup baseLookup;
+
+        protected AbstractClassLoaderLookuper() {
+            try {
+                baseLookup = MethodHandles.lookup();
+            } catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        protected MethodType methodType(Class<?> rtype, Class<?>[] ptypes) {
+            try {
+                return MethodType.methodType(rtype, ptypes);
+            } catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        protected <T> T invoke(ClassLoader classLoader, String methodName, MethodType methodType, Object... arguments) {
+            try {
+                MethodHandles.Lookup lookup = getLookupForClassLoader(classLoader);
+                MethodHandle methodHandle = lookup.findVirtual(ClassLoader.class, methodName, methodType);
```

Comment:
Do we know anything about the performance of these two lines? Should we use a cache?

Reply:
Actually no... but I think our performance tests would complain if it has a huge impact.

Reply:
Fair enough. 🙂